### PR TITLE
Fix RLP encoding for types recursively calling `RlpStream::append`

### LINKED ISSUE #4356: Recursive `RlpStream::append` in `Encodable`/`RlpEncodable` leads to invalid list counts
[Some types](https://github.com/ethcore/parity/blob/8404edb656693270ec1cd3956d204f625d28ec7c/ethcore/src/engines/tendermint/message.rs#L163-L167) implement the `Encodable` trait as follows:

```rust
impl Encodable for Step {
	fn rlp_append(&self, s: &mut RlpStream) {
		s.append(&self.number());
	}
}
```

This seems to generally work, but incorrectly assumes two items got appended when, in fact, only one has been. Isolated way to reproduce the problem:

```rust
    #[test]
    fn message_info_rlp_step() {
        let mut s = RlpStream::new_list(2);
        s.append(&Step::Precommit);
        assert!(!s.is_finished());
    }
```

The stream becomes marked as finished even though only one item has been appended. This is due to the double (recursive) call of `RlpStream::append` - once when appending the item, and another one in the implementation of the trait.

Following through in `RlpStream`:
```rust
	fn append<'a, E>(&'a mut self, value: &E) -> &'a mut Self where E: RlpEncodable {
		self.finished_list = false;
		value.rlp_append(self); // `E` calls append again
		if !self.finished_list {
			self.note_appended(1); // if the list isn't finished, `note_appended` will be called again
		}
		self
	}
```

## Solution:
+ ~Make `RlpStream::append_value` public.~
+ ~The `Encodable` implementations that aren't creating sub-lists should use `append_value` instead of `append`.~
+ Instead of `s.append(&self.to_another_type())` call `self.to_another_type().rlp_append()` to avoid double `RlpStream::append` calls.

## Food for thought:
+ This can't be tested with happy path testing since, should the list be finished after the first `note_appended(1)`, all tests will be green even though two attempts to increase the count has been made. Adding negative tests for `RlpStream::is_finished` would solve that.
* Should `RlpStream` panic :frowning: on overflows? There is no error handling except in `UntrustedRlp`.

### PR DESCRIPTION
**Pull request by maciejhirsz**
Fixes #4356.

### CODE REVIEW

File: ethcore/src/engines/tendermint/message.rs [hunk 1]
```diff
@@ -194,7 +194,7 @@ impl Encodable for ConsensusMessage {
 
 pub fn message_info_rlp(vote_step: &VoteStep, block_hash: Option<BlockHash>) -> Bytes {
 	// TODO: figure out whats wrong with nested list encoding
```
Comment (arkpar) [inline ethcore/src/engines/tendermint/message.rs]: Does this resolve this TODO?

Comment (maciejhirsz) [inline ethcore/src/engines/tendermint/message.rs]: @arkpar yap

**Comment by gavofyork** [comment]
the changes made seem a little subtle - is there any reason these mistakes won't be repeated in the future?

**Comment by maciejhirsz** [comment]
@gavofyork there is more insight into it in #4356. In short - there are use cases when calling `append` inside `rlp_append` are valid - namely when the type implementing `Encodable` creates a list with subitems. 

The long term solution is to refactor how RlpStream works. I'd suggest that we only ever implement `Encodable` for non-list types, and do list by generic array / tuple / Iterator `impl`s, to completely avoid the problem of having to count how many items have been appended inside `RlpStream`.

**Comment by rphmeier** [comment]
> do list by generic array / tuple / Iterator impls

This will be easier said than done as items in RLP lists often have different types. For instance, how would you write a block header as a `Vec<T>` or `Iterator<Item=T>`?

**Comment by maciejhirsz** [comment]
Can use macros to generate generic impl for all tuple sizes up to some limit (20? 32?), such as:

```rust
impl<A, B, C, D...> Encodable for (A, B, C, D...) where
    A: Encodable,
    B: Encodable,
    C: Encodable,
    D: Encodable,
    ...
{ ... }
```

**Pull request merged**